### DISCUSSION TITLE
Decoding Revox B205?

### DISCUSSION THREAD

**Discussion by 3KUdelta** [ukw100/IRMP#80] [Q&A] [answered]
Hello there,
I am trying to decode a 70/80ies HiFi remote from Revox. The remote controls are named B201 and B205. The service manual has the following description:

![Bildschirmfoto 2023-04-24 um 17 07 37](https://user-images.githubusercontent.com/48127424/234039300-cdaa8aa1-adda-44a6-85a1-fd9e15214e4d.png)

Can anyone guide me to declare this in irmpprotocols.h? I tried several versions, but no luck. This was the last one I tried.

Cheers, Marc

```
#define REVOX_START_BIT_PULSE_TIME             450.0e-6                        // 450 usec pulse
#define REVOX_START_BIT_PAUSE_TIME             150.0e-6                        // 150 usec pulse
#define REVOX_0_PULSE_TIME                      150.0e-6                        //  150 usec pulse
#define REVOX_0_PAUSE_TIME                      0                        //  0 usec pause
#define REVOX_1_PULSE_TIME                      300.0e-6                        //  300 usec pulse
#define REVOX_1_PAUSE_TIME                     0                        // 0 usec pause

#define REVOX_FRAME_REPEAT_PAUSE_TIME          4456.0e-6                        // frame repeat after 4460 usec
#define REVOX_ADDRESS_OFFSET                    0                               // skip 1st toggle bit
#define REVOX_ADDRESS_LEN                       4                               // store 7 command bits in address
#define REVOX_COMMAND_OFFSET                    0                               // skip 1st toggle bit + 7 command bits + 2nd toggle bit
#define REVOX_COMMAND_LEN                      6                               // read 7 alternative command bits plus 4 0-bits
#define REVOX_COMPLETE_DATA_LEN                10                               // complete length
#define REVOX_STOP_BIT                          1                               // has stop bit
#define REVOX_LSB                               0                               // MSB...LSB
#define REVOX_FLAGS                             0                               // flags
```

**Comment by ArminJo** [comment]
You are aware of the sample time of IRMP?

**Reply by 3KUdelta** [reply]
I afraid not - since I do not understand where your question leads to.

**Comment by 3KUdelta** [comment]
Hello @ArminJo 
I was reading through the doc at Mikrocontroller.net - sampling rate seems to be 15000/s. Are you telling me that the resolution is not good enough to decode the Revox codes?

**Reply by ukw100** [reply]
Your code as proposed is wrong:
#define REVOX_0_PULSE_TIME                      150.0e-6           //  150 usec pulse
#define REVOX_0_PAUSE_TIME                      0                       //  0 usec pause
#define REVOX_1_PULSE_TIME                      300.0e-6           //  300 usec pulse
#define REVOX_1_PAUSE_TIME                     0                        // 0 usec pause

This should be correct - see picture above:
#define REVOX_0_PULSE_TIME                     15.0e-6              //  15 usec pulse
#define REVOX_0_PAUSE_TIME                     135.0e-6            //  (150 - 15) = 135 usec pause
#define REVOX_1_PULSE_TIME                     15.0e-6              //  15 usec pulse
#define REVOX_1_PAUSE_TIME                     285.0e-6            // (300 - 15) = 285 usec pause

Even if you change the number of interrupts from 15000 to 20000, you have a resolution of 50 usec. This is not enough to detect pulses of 15 usec. Here you need a resolution of ~7 usec (Shannon theorem), which means ~150000 interrupts per second - ten times as much as before.

**Reply by 3KUdelta** [reply]
Thank you indeed for your answer. I highly appreciate it. Since you are the master brain of this, do you see a chance to read and decode these Revox codes with an ESP32?

**Reply by 3KUdelta** [reply]
Further investigations show a VS1838 IR receiver gives the following output (using the example above "Balance L" with output 100110010 on the oscilloscope:

![SDS00009](https://user-images.githubusercontent.com/48127424/234391543-06c222d2-d86a-4b5e-9d6c-3c6abbefd398.png)

For me this does not seem to fit to the specs above. Is the VS1838 resolution also too slow? Revox is using a [TBA 2800 preamplifier IC](https://pl-1.org/getproductfile.axd?id=2569&filename=TBA2800.pdf) in their devices. At least they can handle the pulses via this preamp. It would be great if you could give me some input. 

Cheers, Marc

**Reply by 3KUdelta** [reply]
Yepp, this is the output of the TBA 2800 on pin 8 from a Revox B285. This looks more like the specs (0 and 1 seem to be inverted).

![SDS00011](https://user-images.githubusercontent.com/48127424/234403500-bc684883-0922-45bc-954e-d57c85df1327.jpg)


So, at least I found out that I need such a preamp to properly read the remote -  but there is still my question left: will we be able to decode this code with an ESP32 and IRMP?

**Comment by 3KUdelta** [comment]
Armin, Frank - a simple Yes or No to my question would help me if to further examine the route with IRMP or not.
Cheers, Marc

**Reply by ukw100** [reply, marked as answer]
Sorry, but I don't know if IRMP with 100000 interrupts per second would still work on an ESP32. I have not tried this. So I cannot answer you with Yes or No.

But I rather assume that this will not work. You would have to solve this differently than simple polling in a timer interrupt.

**Reply by 3KUdelta** [reply]
Many thanks. This helps. Once I have found a solution I will post it here as well.
